File: js/database/db-deposit-addrs.ts
import { ChainID } from "../types";
import { MongoClient, Db, Collection, IndexSpecification } from "mongodb";

interface DepositAddressModel {
  agent: string;
  account: number;
  chain: ChainID;
  index: number;
  address: string;
  memo?: string;
  active: boolean;
}

const schema = {
  $jsonSchema: {
    bsonType: "object",
    required: ["agent", "account", "chain", "user", "address", "active"],
    properties: {
      agent: {
        bsonType: "string",
        description: "agentId must be a string and is required",
      },
      account: {
        bsonType: "int",
        description: "account index must be an integer and is required",
      },
      chain: {
        enum: Object.values(ChainID),
        description: "Chain must be a string and is required",
      },
      user: {
        bsonType: "int",
        description: "address index must be an integer and is required",
      },
      address: {
        bsonType: "string",
        description: "address must be a string and is required",
      },
      memo: {
        bsonType: "string",
        description: "memo must be a string and is optional",
      },
      active: {
        bsonType: "bool",
        description: "active must be a boolean and is required",
      },
    },
  },
};

export async function init(db: Db, collectionName: string): Promise<void> {
  // Return if collection exists
  const collections = await db.listCollections({ name: collectionName }).toArray();
  if (collections.length > 0) return;

  // Create collection with schema validation
  await db.createCollection(collectionName, { validator: schema });

  const collection: Collection = db.collection(collectionName);

  // Create unique index
  await collection.createIndex(
    { 
        agent: 1 ,
        account: 1 ,
        chain: 1 ,
        user: 1 
    }, 
    { unique: true });
}

File: custody_service/database/db_deposit_addrs.py
from custody_service.custom_types import ChainId;
from pymongo import ASCENDING
from typing import TypedDict, Optional


class DepositAddressModel(TypedDict):
    agent: str
    acount: int
    chain: ChainId
    index: int
    address: str
    memo: Optional[str]
    active: bool

schema = {
    "$jsonSchema": {
        "bsonType": "object",
        "required": ["agent", "account", "chain", "user", "address", "active"],
        "properties": {
            "agent": {
                "bsonType": "string",
                "description": "agentId must be a string and is required"
            },
            "account": {
                "bsonType": "int",
                "description": "account index must be an integer and is required"
            },
            "chain": {
                "enum": [chain.value for chain in ChainId],
                "description": "Chain must be string and is required"
            },
            # users indexes (user ID)
            "user": {
                "bsonType": "int",
                "description": "address index must be an integer and is required"
            },
            "address": {
                "bsonType": "string",
                "description": "address must be a string and is required"
            },
            "memo": {
                "bsonType": "string",
                "description": "memo must be a string and is optional"
            },
            "active": {
                "bsonType": "bool",
                "description": "active must be a boolean and is required"
            }
        }
    }
}

def init(db, collection_name):
    # return if collection exist
    if collection_name in db.list_collection_names():
        return;
    
    db.create_collection(
        collection_name,
        validator=schema
    )
    collection = db[collection_name]
    collection.create_index(
        [
            ("agent", ASCENDING), 
            ("account", ASCENDING), 
            ("chain", ASCENDING), 
            ("user", ASCENDING)
        ], 
        unique=True
    )